src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java

```
@@ -312,6 +391,205 @@ public void processElement(final ProcessContext context) throws Exception {
+                                ++metrics.UNPAIRED_READS_EXAMINED;
+                            } else {
+                                ++metrics.READ_PAIRS_EXAMINED;
+                            }
+
+                            if (record.isDuplicate()) {
+                                if (!record.isPaired() || record.mateIsUnmapped()) {
+                                    ++metrics.UNPAIRED_READ_DUPLICATES;
+                                } else {
+                                    ++metrics.READ_PAIR_DUPLICATES;
+                                }
+                            }
+                            if (record.hasAttribute(OPTICAL_DUPLICATE_ATTRIBUTE_NAME) &&
+                                record.getAttributeAsInteger(OPTICAL_DUPLICATE_ATTRIBUTE_NAME) == 1) {
+                                ++metrics.READ_PAIR_OPTICAL_DUPLICATES;
+                            }
+                            final KV<String, DuplicationMetrics> kv = KV.of(library, metrics);
+                            context.output(kv);
+                        }
+                    });
+        }
+
+
+        public static class CombineMetricsFn implements SerializableFunction<Iterable<DuplicationMetrics>, DuplicationMetrics> {
+            private static final long serialVersionUID = 1l;
+
+            @Override
+            public DuplicationMetrics apply(Iterable<DuplicationMetrics> input) {
+                DuplicationMetrics metricsSum = new DuplicationMetrics();
+                int num = 0;
+                for (final DuplicationMetrics metrics : input) {
+                    if (metricsSum.LIBRARY == null) {
+                        metricsSum.LIBRARY = metrics.LIBRARY;
+                    }
+                    // This should never happen, as we grouped by key using library as the key.
+                    if (!metricsSum.LIBRARY.equals(metrics.LIBRARY)) {
+                        return null;
```

If this is an error condition it should throw an exception instead I think.